lib/widgets/actions.dart

```diff
@@ -16,101 +16,171 @@ import '../model/narrow.dart';
-  Narrow narrow,
-  bool useLegacy, // TODO(server-6)
-) async {
-  final store = PerAccountStoreWidget.of(context);
-  final connection = store.connection;
-  if (useLegacy) {
-    return await _legacyMarkNarrowAsRead(context, narrow);
-  }
-
-  // Compare web's `mark_all_as_read` in web/src/unread_ops.js
-  // and zulip-mobile's `markAsUnreadFromMessage` in src/action-sheets/index.js .
-  final zulipLocalizations = ZulipLocalizations.of(context);
-  final scaffoldMessenger = ScaffoldMessenger.of(context);
-  // Use [AnchorCode.oldest], because [AnchorCode.firstUnread]
-  // will be the oldest non-muted unread message, which would
-  // result in muted unreads older than the first unread not
-  // being processed.
-  Anchor anchor = AnchorCode.oldest;
-  int responseCount = 0;
-  int updatedCount = 0;
-
-  // Include `is:unread` in the narrow.  That has a database index, so
-  // this can be an important optimization in narrows with a lot of history.
-  // The server applies the same optimization within the (deprecated)
-  // specialized endpoints for marking messages as read; see
-  // `do_mark_stream_messages_as_read` in `zulip:zerver/actions/message_flags.py`.
-  final apiNarrow = narrow.apiEncode()..add(ApiNarrowIsUnread());
+Future<void> markNarrowAsRead(BuildContext context, Narrow narrow) async {
+  try {
+    final store = PerAccountStoreWidget.of(context);
+    final connection = store.connection;
+    final useLegacy = connection.zulipFeatureLevel! < 155; // TODO(server-6)
+    if (useLegacy) {
+      await _legacyMarkNarrowAsRead(context, narrow);
+      return;
+    }
 
-  while (true) {
-    final result = await updateMessageFlagsForNarrow(connection,
-      anchor: anchor,
+    // Compare web's `mark_all_as_read` in web/src/unread_ops.js
+    // and zulip-mobile's `markAsUnreadFromMessage` in src/action-sheets/index.js .
+    final zulipLocalizations = ZulipLocalizations.of(context);
+    final didPass = await updateMessageFlagsStartingFromAnchor(
+      context: context,
+      // Include `is:unread` in the narrow.  That has a database index, so
+      // this can be an important optimization in narrows with a lot of history.
+      // The server applies the same optimization within the (deprecated)
+      // specialized endpoints for marking messages as read; see
+      // `do_mark_stream_messages_as_read` in `zulip:zerver/actions/message_flags.py`.
+      apiNarrow: narrow.apiEncode()..add(ApiNarrowIsUnread()),
+      // Use [AnchorCode.oldest], because [AnchorCode.firstUnread]
+      // will be the oldest non-muted unread message, which would
+      // result in muted unreads older than the first unread not
+      // being processed.
+      startingAnchor: AnchorCode.oldest,
       // [AnchorCode.oldest] is an anchor ID lower than any valid
       // message ID; and follow-up requests will have already
       // processed the anchor ID, so we just want this to be
       // unconditionally false.
       includeAnchor: false,
```

Comment:
The reasoning in the second half of this comment applies just as well on markNarrowAsUnreadFromMessage, or any other caller of updateMessageFlagsStartingFromAnchor, as it does on markNarrowAsRead. It's a fact about how the loop works, rather than about the value of `startingAnchor`.

So in the loop we should make `includeAnchor` be false after the first iteration, though the first iteration will depend on the caller.